pkg/urbit/noun/retrieve.c

```diff
@@ -1061,22 +1061,30 @@ u3r_met(c3_y    a_y,
         */
         c3_w bif_w, col_w;
 
+        if ( gal_w > ((UINT32_MAX - 35) >> 5) ) {
+          return u3m_bail(c3__fail);
+        }
+
         col_w = c3_bits_word(daz_w);
         bif_w = col_w + (gal_w << 5);
 
         return (bif_w + ((1 << a_y) - 1)) >> a_y;
       }
-      case 3: {
-        return  (gal_w << 2)
-              + ((daz_w >> 24) ? 4 : (daz_w >> 16) ? 3 : (daz_w >> 8) ? 2 : 1);
-      }
-      case 4: {
-        return  (gal_w << 1)
-              + ((daz_w >> 16) ? 2 : 1);
-      }
+
+      STATIC_ASSERT((UINT32_MAX > ((c3_d)u3a_maximum << 2)),
+                    "met overflow");
+
+      case 3: return (gal_w << 2) + ((c3_bits_word(daz_w) +  7) >> 3);
+
+      case 4: return (gal_w << 1) + ((c3_bits_word(daz_w) + 15) >> 4);
+
       default: {
         c3_y gow_y = (a_y - 5);
 
+        if ( gow_y > 31 ) {
+          return u3m_bail(c3__fail);
+        }
```

Comment:
On second thought, adding this kind of check feels like a slippery slope. We'll never hit it in practice through a jet (all user-proved bloqs must be checked to be less than 32 already), so it's only guarding against a statically invalid call. Such invariants already abound, and this one is particularly absurd ("i want the length of this atom in terms of `1 << 37`- bit chunks"). Removing in favor of documenting the invariant.